dds_web/database/models.py

```diff
@@ -370,40 +375,42 @@ def verify_reset_token(token):
         return User.query.get(user_id)
 
     # 2FA related
-    def gen_otp_secret(self):
-        """Generate random base 32 for otp secret."""
-        return pyotp.random_base32()
-
-    @property
-    def otp_secret(self):
-        """Get OTP secret for user if the 2fa has not already been setup."""
-        if self.has_2fa:
-            return None
-        return self._otp_secret
-
-    @otp_secret.setter
-    def otp_secret(self, otp_secret):
-        """Set new otp secret."""
-        if not otp_secret:
-            otp_secret = self.gen_otp_secret()
-        self._otp_secret = otp_secret
-
-    def set_2fa_seen(self):
-        """Renew the otp secret -- one should only be setup once."""
-        self.has_2fa = True
-
-    def totp_uri(self):
-        """Get uri for user otp_secret if the 2fa has not already been setup."""
-        if self.has_2fa:
-            return None
-        return pyotp.totp.TOTP(self.otp_secret).provisioning_uri(
-            name=self.username, issuer_name="Data Delivery System"
+    def generate_HOTP_token(self):
+        """Generate a one time password, e.g. to be sent by email.
+
+        Counter is incremented before generating token which invalidates any previous token.
+        The time when it was issued is recorded to put an expiration time on the token.
+
+        """
+        self.hotp_counter += 1
+        self.hotp_requested_time = dds_web.utils.current_time()
+        db.session.commit()
+        flask.current_app.logger.info(
+            f"Incremented counter to: {self.hotp_counter} and saved time: {self.hotp_requested_time}"
         )
 
-    def verify_totp(self, token):
-        """Verify the otp token."""
-        totp = pyotp.TOTP(self.otp_secret)
-        return totp.verify(token, valid_window=1)
+        hotp = twofactor_hotp.HOTP(self.hotp_secret, 8, hashes.SHA512())
+        return hotp.generate(self.hotp_counter)
+
+    def verify_HOTP(self, token):
+        """Verify the HOTP token.
+
+        raises AuthenticationError if token is invalid or has expired (older than 1 hour).
+        If the token is valid, the counter is incremented, to prohibit re-use.
+        """
+        hotp = twofactor_hotp.HOTP(self.hotp_secret, 8, hashes.SHA512())
+        timediff = dds_web.utils.current_time() - self.hotp_requested_time
+        if timediff > datetime.timedelta(minutes=15):
+            raise AuthenticationError("One-time authentication code has expired.")
+
+        try:
+            hotp.verify(token, self.hotp_counter)
+        except twofactor_InvalidToken:
+            raise AuthenticationError("Invalid one-time authentication code.")
+
+        # Token verified, increment counter to prohibit re-use
+        self.hotp_counter += 1
+        db.session.commit()
```

Comment:
I think that lines 412 and 413 are not relevant and can be removed. Generate_HOTP_token already increments the counter before generating. Counter increment on line 412 doesn't happen on invalid one-time auth code and happens on successful verification. So on successful verification the counter is incremented twice.

Reply:
Without this, one could use the same HOTP token twice, as long as a new one has not been requested. While it might not be that easy to actually do it. It's very easy to use this to prevent it completely.

Reply:
Yes. Keep this.